A32-xds-circuit-breaking.md

```diff
@@ -0,0 +1,147 @@
+# gRPC xDS circuit breaking
+
+* Author(s): Chengyuan Zhang (voidzcy)
+* Approver: markdroth
+* Status: In Review
+* Implemented in:
+* Last updated: 2020-07-20
+* Discussion at: https://groups.google.com/g/grpc-io/c/NEx70p8mcjg
+
+
+## Abstract
+
+Adopt Envoy's [circuit breaking](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/circuit_breaking#circuit-breaking) 
+component that comes with the xDS protocol to
+fail network operations quickly and apply back pressure downstream as soon as
+possible. This allows gRPC clients to received distributed traffic limit 
+configurations that prevent cascaded service overload failures. 
+
+## Background
+
+Envoy supports configurable circuit breaking settings on per upstream cluster 
+and per priority basis to allow different components of the distributed system
+to be tuned independently. Since gRPC xDS does not support [priority based
+routing](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/http/http_routing#arch-overview-http-routing-priority), 
+requests to a single upstream cluster conform to the single circuit breaking
+configuration.
+
+Envoy handles HTTP traffic with [connection pooling](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/connection_pooling#connection-pooling),
+and many circuit breaking parameters are configured for
+connection/connection pool management. Since each gRPC client creates at most
+one HTTP/2 connection to a single upstream endpoint address, those
```

Comment:
Naive question; I had always thought that gRPC reserves the right to back a single "virtual" channel with multiple connections. Do you want to provide future proofing by reserving the right to add these multi-connection parameters in the future?

Reply:
Just from the design of [`Subchannel`](https://github.com/grpc/grpc-java/blob/e19bdf33b1d647c282169bc32a98362d53dd340b/api/src/main/java/io/grpc/LoadBalancer.java#L1201-L1217), it is intended to be a single connection at most. Although Java and Go does not have Subchannel sharing mechanism (C-core does have), we do have considerations trying to avoid cases that could potentially create more than one Subchannels to the same endpoint (e.g., [the RouteAction support design](https://github.com/grpc/proposal/pull/192) is trying to avoid having multiple CDS/EDS policies for the same cluster, and therefore a single leaf policy that creates actual connections to endpoints in the cluster). Based on that, I think it is safe to say we will not have more than one connections per endpoint. 

> Backing a single "virtual" channel (aka, Subchannel) with multiple connections.

That may be too far to be mentioned, unless the abstraction for connection/transport layer in gRPC is redesigned.

Reply:
Ack. It seems to be an implementation limitation but not some sort of in-principle limit.

Reply:
I don't think anything in this design stops us from adding support for these parameters in the future if/when it becomes necessary.

For example, it's possible that in the future, we might add some new type of LB policy that would allow multiple connections to address the MAX_CONCURRENT_STREAMS limit from the server.  If/when we do that, we can reevaluate whether supporting this kind of circuit breaker makes sense.